### PR TITLE
SNOW-878073 Fix of retry algorithm

### PR DESCRIPTION
**Pull request by sfc-gh-pmotacki**
### Description
Please explain the changes you made here.

### Checklist
- [ ] Code compiles correctly
- [ ] Run ``make fmt`` to fix inconsistent formats
- [ ] Run ``make lint`` to get lint errors and fix all of them
- [ ] Created tests which fail without the change (if possible)
- [ ] All tests passing
- [ ] Extended the README / documentation, if necessary

### CODE REVIEW

File: retry.go [hunk 1]
```diff
@@ -208,24 +208,19 @@ func isQueryRequest(url *url.URL) bool {
...
-	jitteredSleepTime := math.Pow(2, float64(attempt)) + jitterAmount
+	jitterAmount := w.getJitter(currWaitTime)
+	jitteredSleepTime := chooseRandomInRange(currWaitTime+jitterAmount, math.Pow(2, float64(attempt))+jitterAmount)
 	return jitteredSleepTime
 }
 
-func (w *waitAlgo) getJitter(currWaitTime float64, jitterPercentage float64) float64 {
-	multiplicationFactor := chooseRandomFromValues(w.random, []int{-1, 1}) // random int from [-1, 1]
-	jitterAmount := jitterPercentage * currWaitTime * float64(multiplicationFactor)
+func (w *waitAlgo) getJitter(currWaitTime float64) float64 {
```
Comment (sfc-gh-dheyman-1) [inline retry.go:216]: i would add a test that checks if returned value is is specific range for couple input values. E.g. if currWaitTime is 2 then jitter should be in range -1:1 etc.

File: retry.go [hunk 2]
```diff
@@ -208,24 +208,19 @@ func isQueryRequest(url *url.URL) bool {
...
 	w.mutex.Lock()
 	defer w.mutex.Unlock()
-	var jitterPercentage = 0.5
-	if attempt < 2 {
-		jitterPercentage = 0.25 // to ensure there will be sleep time increase between attempts
-	}
-	jitterAmount := w.getJitter(currWaitTime, jitterPercentage)
-	jitteredSleepTime := math.Pow(2, float64(attempt)) + jitterAmount
+	jitterAmount := w.getJitter(currWaitTime)
+	jitteredSleepTime := chooseRandomInRange(currWaitTime+jitterAmount, math.Pow(2, float64(attempt))+jitterAmount)
```
Comment (sfc-gh-pfus) [inline retry.go:212]: `chooseRandomFromRange`?

File: retry.go [hunk 3]
```diff
@@ -208,24 +208,19 @@ func isQueryRequest(url *url.URL) bool {
...
 	return jitteredSleepTime
 }
 
-func (w *waitAlgo) getJitter(currWaitTime float64, jitterPercentage float64) float64 {
-	multiplicationFactor := chooseRandomFromValues(w.random, []int{-1, 1}) // random int from [-1, 1]
-	jitterAmount := jitterPercentage * currWaitTime * float64(multiplicationFactor)
+func (w *waitAlgo) getJitter(currWaitTime float64) float64 {
+	multiplicationFactor := chooseRandomInRange(-1, 1) // random int from [-1, 1]
```
Comment (sfc-gh-pfus) [inline retry.go:217]: Do we need this comment?

File: retry.go [hunk 4]
```diff
@@ -208,24 +208,19 @@ func isQueryRequest(url *url.URL) bool {
...
+	jitterAmount := 0.5 * currWaitTime * multiplicationFactor
 	return jitterAmount
 }
 
-func chooseRandomFromValues[T any](random *rand.Rand, arr []T) T {
-	valIdx := random.Intn(len(arr))
-	return arr[valIdx]
+func chooseRandomInRange(min float64, max float64) float64 {
```
Comment (sfc-gh-pfus) [inline retry.go:222]: Maybe it would be better to move it to `util.go`?

File: retry.go (hunk 1 above)
Comment (sfc-gh-pmotacki) [reply retry.go:216]: I've added test.

File: retry.go (hunk 2 above)
Comment (sfc-gh-pmotacki) [reply retry.go:212]: ok

File: retry.go [hunk 5]
```diff
@@ -208,24 +208,19 @@ func isQueryRequest(url *url.URL) bool {
...
+	return rand.Float64()*(max-min) + min
```
[the remark this answers is not in the document]
Comment (sfc-gh-pmotacki) [reply retry.go:223]: I added the panic error

File: retry.go (hunk 4 above)
Comment (sfc-gh-pmotacki) [reply retry.go]: ok

**Review by sfc-gh-dheyman-1 — APPROVED** [review]

**Comment by sfc-gh-pmotacki** [comment]
I have read the CLA Document and I hereby sign the CLA

File: retry_test.go [hunk 6]
```diff
@@ -533,18 +532,78 @@ func TestIsRetryable(t *testing.T) {
...
+			minSleepTime: 16.178897,
+			maxSleepTime: 1048592.178897,
+		},
 	}
 
-	for i := 0; i < 9; i++ {
-		if retryTimes[i] >= retryTimes[i+1] {
-			log.Fatalf("expected consequent values to be greater than previous ones; array: %v", retryTimes)
-		}
+	for _, tc := range tcs {
```
Comment (sfc-gh-pfus) [inline retry_test.go:605]: Can we have `t.Run` for each test case?

File: retry_test.go [hunk 7]
```diff
@@ -533,18 +532,78 @@ func TestIsRetryable(t *testing.T) {
...
+		result := defaultWaitAlgo.calculateWaitBeforeRetry(tc.attempt, tc.currWaitTime)
+		assertTrueF(t, result > tc.minSleepTime && result < tc.maxSleepTime, fmt.Sprintf("sleep time should be between min: %f and max: %f", tc.minSleepTime, tc.maxSleepTime))
```
Comment (sfc-gh-pfus) [inline retry_test.go:607]: Can we add `assertBetweenE`? And error is better here?

Comment (sfc-gh-pmotacki) [reply retry_test.go:607]: yes. Good idea

File: retry_test.go (hunk 6 above)
Comment (sfc-gh-pmotacki) [reply retry_test.go:605]: Ok

File: retry_test.go [hunk 8]
```diff
@@ -533,18 +532,81 @@ func TestIsRetryable(t *testing.T) {
...
+			maxSleepTime: 1048592.178897,
+		},
 	}
 
-	for i := 0; i < 9; i++ {
-		if retryTimes[i] >= retryTimes[i+1] {
-			log.Fatalf("expected consequent values to be greater than previous ones; array: %v", retryTimes)
-		}
+	for _, tc := range tcs {
+		t.Run(fmt.Sprintf("attmep: %v", tc.attempt), func(t *testing.T) {
```
Comment (sfc-gh-pfus) [inline retry_test.go:607]: attempt

File: retry_test.go [hunk 9]
```diff
@@ -533,18 +532,81 @@ func TestIsRetryable(t *testing.T) {
...
-		retryTimes[i] = resultTime
-		inputTime = resultTime
+func TestCalculateRetryWait(t *testing.T) {
+
+	tcs := []struct {
+		attempt      int
+		currWaitTime float64
+		minSleepTime float64
+		maxSleepTime float64
+	}{ // test for randomly selected attempt and currWaitTime values
```
Comment (sfc-gh-pfus) [inline retry_test.go:542]: Maybe the comment to line 536 - this one is blank and awaits for some content ;)

**Review by sfc-gh-pfus — APPROVED** [review]

**Pull request merged**